course-definition.yml

```diff
@@ -557,3 +565,326 @@ stages:
+      [extension handshake](https://www.bittorrent.org/beps/bep_0010.html#handshake-message). This message includes a dictionary that maps extension names to IDs.
+
+      Extension messages follow the standard BitTorrent message format:
+
+      - message length prefix (4 bytes)
+      - message id (1 byte)
+        - This will be 20 for all messages implemented by extensions
+      - payload (variable size)
+
+      The payload will be structured as follows:
+
+      - extension message id (1 byte)
+        - This will be 0 for the extension handshake
+      - bencoded dictionary (variable size)
+        - This will contain a key "m" with another dictionary as its value.
+        - The inner dictionary maps supported extension names to their corresponding message IDs.
+        - For example, the inner dictionary contents might be `{"ut_metadata": 1, "ut_pex": 2}`, indicating that the peer supports the "ut_metadata" and "ut_pex" extensions with IDs 1 and 2 respectively.
+
+      The extension name we're interested in is "ut_metadata", which corresponds to the [metadata extension](https://www.bittorrent.org/beps/bep_0009.html#ut_metadata).
+
+      ### Tests
+
+      In this stage, you'll print out the message ID your peer uses for the metadata extension (the value of "ut_metadata" key in the inner dictionary)
+
+      Here's how the tester will execute your program:
+
+      ```bash
+      $ ./your_bittorrent.sh magnet_handshake <magnet-link>
+      ```
+
+      and here's the output it expects:
+
+      ```
+      Peer ID: 0102030405060708090a0b0c0d0e0f1011121314
+      Peer Metadata Extension ID: 123
+      ```
+
+      ### Notes
+
+      - The value of `Peer Metadata Extension ID` will be randomly generated, you'll need to fetch it from the dictionary in the handshake message.
+      - Extension IDs need to be stored for every peer, as different peers may use different IDs for the same extension.
+      - Client ignore any unrecognized extension names.
+      - To avoid collisions, extension names are prefixed with two characters identifying the client that introduced the extension (e.g., "**lt**" for libtorrent or "**ut**" for μTorrent)
+
+    marketing_md: |-
+      In this stage, you'll receive the list of extensions your peer supports.
+
+  - slug: "xi4"
+    primary_extension_slug: "magnet-links"
+    name: "Send Extension Handshake"
+    difficulty: easy
+    description_md: |-
+      In this stage, you'll send back a list of extensions that your client supports.
+
+      ### Extension handshake (part 2)
+
+      Similar to how your client received an extension handshake message from your peer in the previous stage,
+      your client will need to send an extension handshake message to the peer too.
+
+      This message must be sent before receiving the extension handshake (if both clients waited for the other to send a handshake, they'd be stuck forever).
```

Comment:
@sarp this is one part I wasn't sure about (related to the question I tagged you on Slack about)

Reply:
Yep, showing Send Extension stage before Receive Extension stage should do the trick?

Do we need to over emphasize sending before receiving in the description? If we order stages accordingly, I bet people will implement it that way.

Reply:
Yep, true - will amend